src/hashes/interface.ts

```diff
@@ -37,14 +39,30 @@ export interface MultihashDigest<Code extends number = number> {
  */
 export interface MultihashHasher<Code extends number = number> {
   /**
-   * Takes binary `input` and returns it (multi) hash digest. Return value is
+   * Takes binary `input` and returns it multihash digest. Return value is
+   * either promise of a digest or a digest. This way general use can `await`
+   * while performance critical code may asses return value to decide whether
+   * await is needed.
+   *
+   * @param {Uint8Array} input
+   */
+  digest(input: Uint8Array): Await<MultihashDigest>
+
+  /**
+   * Takes binary `input` and returns it plain hash digest. Return value is
    * either promise of a digest or a digest. This way general use can `await`
    * while performance critical code may asses return value to decide whether
    * await is needed.
    *
+   * This is distinct from `digest()` which returns a multihash (prefixed)
+   * digest for this hasher as it only returns the encoded bytes that may
+   * otherwise be obtained from `digest().digest`. Only use `encode()` if you
+   * need to use this hasher as a standard hash digest generator; multihashes
+   * should otherwise be preferred.
+   *
    * @param {Uint8Array} input
    */
-  digest(input: Uint8Array): Promise<MultihashDigest> | MultihashDigest
+  encode(input: Uint8Array): Await<Uint8Array>
```

Comment:
I think this would be a mistake as suddenly passing `func({ digest }) ` will no longer satisfy requirements of the function even if it does not care about encode.

Proper way to address underlying problem is not to constraint  interface further by increasing it’s surface but rather to make our type annotations more precise. E.g it is ok to say that our multihashers implement both MultihashHasher interface along with some Encoder interface.

We can even define new interfaces here here that extend both to reference it across the board. We really should avoid bloating the interface further so that it’s easier to be a valid implementation of. As shown in example it can actually be a TS breaking change, as some code may become invalid.

P.S. I’m kind of regretting name field which in most code is unecessary contstraint.